openstack/identity/v3/applicationcredentials/requests.go

```diff
@@ -0,0 +1,95 @@
+package applicationcredentials
+
+import (
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/pagination"
+)
+
+// ListOptsBuilder allows extensions to add additional parameters to
+// the List request
+type ListOptsBuilder interface {
+	ToApplicationCredentialListQuery() (string, error)
+}
+
+// ListOpts provides options to filter the List results.
+type ListOpts struct {
+	// Name filters the response by an application credential name
+	Name string `json:"name"`
+}
+
+// ToApplicationCredentialListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToApplicationCredentialListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// List enumerates the ApplicationCredentials to which the current token has access.
+func List(client *gophercloud.ServiceClient, userID string, opts ListOptsBuilder) pagination.Pager {
+	url := listURL(client, userID)
+	if opts != nil {
+		query, err := opts.ToApplicationCredentialListQuery()
+		if err != nil {
+			return pagination.Pager{Err: err}
+		}
+		url += query
+	}
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return ApplicationCredentialPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// Get retrieves details on a single user, by ID.
+func Get(client *gophercloud.ServiceClient, userID string, id string) (r GetResult) {
+	_, r.Err = client.Get(getURL(client, userID, id), &r.Body, nil)
+	return
+}
+
+// CreateOptsBuilder allows extensions to add additional parameters to
+// the Create request.
+type CreateOptsBuilder interface {
+	ToApplicationCredentialCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts provides options used to create an application credential.
+type CreateOpts struct {
```

Comment:
It looks like this might support arbitrary extra fields: https://github.com/openstack/keystone/blob/ef87a202127f0ac84b857716e72595ce75058378/keystone/application_credential/schema.py#L49

This is common with the Keystone/Identity service. Basically it allows you to add _any_ additional field to the request and it will be stored in an `extra` db column.

See here for an example of how to implement this:

* https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/users/requests.go#L126
* https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/users/requests.go#L142

If this feature works for application credentials, a similar implementation can be done here. However, that's optional, especially for an initial commit.

Reply:
Looks like it doesn't work, at least in rocky:

```json
2019/02/14 09:12:45 [DEBUG] OpenStack Request Body: {
  "application_credential": {
    "description": "Swift technical user",
    "expires_at": "2019-03-09T12:58:50Z",
    "foo": "bar",
    "name": "swift",
    "roles": [
      {
        "name": "swiftoperator"
      }
    ],
    "unrestricted": false
  }
}
2019/02/14 09:12:47 [DEBUG] OpenStack Response Code: 400
2019/02/14 09:12:47 [DEBUG] OpenStack Response Headers:
Content-Length: 131
Content-Type: application/json
Date: Thu, 14 Feb 2019 08:12:47 GMT
Server: nginx/1.15.3
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: X-Auth-Token
X-Openstack-Request-Id: req-48f7bcdf-8f71-4003-af22-38667c601990
2019/02/14 09:12:47 [DEBUG] OpenStack Response Body: {
  "error": {
    "code": 400,
    "message": "u'foo' is an invalid keyword argument for ApplicationCredentialModel",
    "title": "Bad Request"
  }
}
```

Reply:
Thank you for looking into it.